File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLength=0;
        int left=0;
        unordered_map<char,int>lastseen;
        for(int right=0;right<s.length();right++){
            char c=s[right];
            if(lastseen.find(c)!=lastseen.end() && lastseen[c]>=left){
                left=lastseen[c]+1;
            }
            maxLength=max(maxLength,right-left+1);
            lastseen[c]=right;
        }
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq = defaultdict(int)
        left = 0
        ans = 0
        for right in range(len(s)):
            char = s[right]
            freq[char] = freq.get(char, 0) + 1

            while freq[char] > 1:
                freq[s[left]] -= 1
                left += 1
            ans = max(ans, right - left + 1)

        return ans
